db_migrator--0.9.0.sql

```diff
@@ -20,25 +20,179 @@ CREATE FUNCTION materialize_foreign_table(
+                  WHEN v_type = 'LIST' THEN format(
+                     'FOR VALUES IN (%s)', 
+                     v_values
+                  )
+                  WHEN v_type = 'RANGE' THEN format(
+                     'FOR VALUES FROM (%s) TO (%s)',
+                     coalesce(v_lowervalues, 'MINVALUE'), v_values
+                  )
+                  WHEN v_type = 'HASH' THEN format(
+                     'FOR VALUES WITH (modulus %s, remainder %s)', 
+                     v_lastpos, v_pos - 1
+                  )
+               END
+            );
+
+            /* create subpartition */
+            OPEN cur_subpartitions FOR EXECUTE format($stmt$
+               SELECT schema, table_name, partition_name, subpartition_name,
+                     type, expression, position, values, is_default
+                  FROM %I.subpartitions 
+                  WHERE table_name = $1 AND partition_name = $2 AND migrate
+                  ORDER BY position $stmt$, pgstage_schema)               
+               USING table_name, v_partition;
+
+            FETCH cur_subpartitions INTO
+               v_schema, v_table, v_partition, v_subpartition,
+               v_type, v_exp, v_pos, v_subvalues, v_default;
+
+            with_subpartitions := FOUND;
+
+            IF with_subpartitions THEN
+               IF (v_type = 'HASH') THEN
+                  /* retrieve the last subpartition definition */
+                  FETCH LAST FROM cur_subpartitions INTO
+                     v_schema, v_table, v_partition, v_subpartition,
+                     v_type, v_exp, v_lastpos, v_subvalues, v_default;
+
+                  /* go back to the first row */
+                  FETCH FIRST FROM cur_subpartitions INTO
+                     v_schema, v_table, v_partition, v_subpartition,
+                     v_type, v_exp, v_pos, v_subvalues, v_default;
+               END IF;
+
+               stmt := format('%s PARTITION BY %s (%s)', stmt, v_type, v_exp);
+            END IF;
+            EXECUTE stmt;
+
+            IF with_subpartitions THEN
+               LOOP
+                  stmt := format(
+                     'CREATE TABLE %1$I.%3$I PARTITION OF %1$I.%2$I %4$s',
+                     v_schema, v_partition, v_subpartition,
+                     CASE
+                        WHEN v_default THEN 'DEFAULT'
+                        WHEN v_type = 'LIST' THEN format(
+                           'FOR VALUES IN (%s)', 
+                           v_values
+                        )
+                        WHEN v_type = 'RANGE' THEN format(
+                           'FOR VALUES FROM (%s) TO (%s)',
+                           coalesce(v_sublowvalues, 'MINVALUE'), v_subvalues
+                        )
+                        WHEN v_type = 'HASH' THEN format(
+                           'FOR VALUES WITH (modulus %s, remainder %s)', 
+                           v_lastpos, v_pos - 1
+                        )
```

Comment:
Again, an error message for other value would be good.

Reply:
Same here, I supposed constraints do the job

```sql
CONSTRAINT subpartition_chk_type 
         CHECK (type IN ('LIST', 'RANGE', 'HASH'))
```